### PR TITLE
Basic batch gradients workflow fixes

### PR DESCRIPTION
**Pull request by cnaize**
Hi,
Thank you so much for your lib, especially for pure Lua 5.1 support!

I'm trying to use it in my own project and faced some issues.
I'm newbie in neural networks, but here are my fixes.

Functions:
- lnn.forwardpass()
- lnn.adjust.adjustfromgradient()
- lnn.adjust.basic.returngradient()
- lnn.data.exportdata()

Changed:
- added support for Softmax activation function
- fixed support for different layers sizes
- fixed file cleanup before export
- improved code readability (but slower performance?)

### CODE REVIEW

**Comment by x-xxoa** [comment]
Hi there!

Thank you for contributing to this repository. Your improvements are highly appreciated and I truly appreciate the time and effort you've dedicated to enhancing the code.

I wanted to reach out and provide an update on the projects direction. In the upcoming version, v1.3.00-unstable, I've decided to start from scratch when development on v1.3.00 began. This decision is to add support for new features such as support for different NN architectures, layers, regularizers, etc. and to address issues with the library.

I want to emphasize that your efforts have not gone unnoticed. I found that your improvements to code readability valuable and could greatly improve the overall quality of the project.

Please know that the decision to start from scratch is not a reflection of the quality of your work, but rather driven by long term goals to make the library more robust.

I would be thrilled if you would like to collaborate on v1.3.00. Your insights and ideas can have a significant impact on the new version as so far this is just a project I've been working on in my free time (well i try but focusing is not something im good at like 90% of the time).

While I'm happy and excited about your contribution, I feel guilty that your time and effort that you put into the code is going to be replaced. Please know it's not a reflection of the quality of your work, but rather long term goals for the project.

Once again, thank you for the contribution and please reach out to me if you have any questions or want to discuss anything further.

Now regarding the code itself: On line 397, I noticed instead of looping through `nextlayer` and setting `nextlayer[i]` to `outsoftmax[i]`, we could simplify it by setting `nextlayer` to `outsoftmax`. If there is a reason for the loop that I'm not aware of, I apologize.

Other than that, everything else looks good though!

Best regards,
xxoa

**Comment by cnaize** [comment]
Thank you for your detailed answer!

Greate to know that you are already implementing v1.3.00-unstable. I'm gonna watch and wait for stable version :)

I really like the idea to make pure Lua 5.1 support, because I'm already using it from restricted environments, like https://github.com/yuin/gopher-lua.

Regarding to your point about line 397.
Yes, it looks like overhead, but `nextlayer` in scope of the function is just local variable. I can overwrite it, but original layer will stay unchanged.

Anyway, it might be good idea to accept my pull request, if someone is trying your lib right now, as me.
